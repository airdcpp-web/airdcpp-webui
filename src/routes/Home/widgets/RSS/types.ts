import * as UI from '@/types/ui';

export type FeedNode<Attributes extends object = UI.EmptyObject> =
  | string
  | {
      text: string;
      attr: Attributes;
    };

export interface FeedItem {
  title?: FeedNode;
  updated?: FeedNode;
  pubDate?: FeedNode;
  guid?: FeedNode;
  id?: FeedNode;
  link?: FeedNode<{
    href: string;
  }>;
}

export interface RawFeedData {
  rss?: {
    channel?: {
      item?: FeedItem[];
    };
  };
  feed?: {
    entry?: FeedItem[];
  };
}
