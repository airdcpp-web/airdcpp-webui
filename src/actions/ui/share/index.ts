export * from './ShareActions';
export * from './ShareExcludeActions';
export * from './ShareProfileActions';
export * from './ShareRootActions';
