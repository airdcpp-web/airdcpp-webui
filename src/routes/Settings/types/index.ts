import * as API from '@/types/api';
import * as UI from '@/types/ui';

export type SettingSectionLayoutProps = Pick<UI.RouteComponentProps, 'location'> &
  React.PropsWithChildren<{
    contentClassname: string;
    selectedRootMenuItem: RootSectionType;
    selectedChildMenuItem: ChildSectionType;
    menu: {
      childMenuItems: React.ReactNode[];
      childAdvancedMenuItems?: React.ReactNode[];
      rootMenuItems: React.ReactNode[];
    };
    settingsT: UI.ModuleTranslator;
    message: React.ReactNode;
    getSaveButton: (className?: string) => React.ReactNode;
  }>;

export type SettingPageProps = React.PropsWithChildren<{
  settingsT: UI.ModuleTranslator;
  moduleT: UI.ModuleTranslator;
}>;

export interface SectionBase {
  url: string;
  access?: API.AccessEnum;
  title: string;
}

export interface ChildSectionType extends SectionBase {
  noSave?: boolean;
  local?: boolean;
  component: React.ComponentType<SettingPageProps>;
}

export interface RootSectionType extends SectionBase {
  menuItems: ChildSectionType[];
  advancedMenuItems?: ChildSectionType[];
  icon: string;
  component: React.ComponentType;
}
