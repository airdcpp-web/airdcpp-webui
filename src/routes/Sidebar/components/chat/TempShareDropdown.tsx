import * as React from 'react';
import cx from 'classnames';

import SectionedDropdown from '@/components/semantic/SectionedDropdown';
import MenuSection from '@/components/semantic/MenuSection';
import MenuItemLink from '@/components/semantic/MenuItemLink';

import * as API from '@/types/api';
import * as UI from '@/types/ui';

import DataProviderDecorator, {
  DataProviderDecoratorChildProps,
} from '@/decorators/DataProviderDecorator';
import ShareConstants from '@/constants/ShareConstants';
import NotificationActions from '@/actions/NotificationActions';
import IconConstants from '@/constants/IconConstants';
import FileIcon from '@/components/icon/FileIcon';

import { translate, toI18nKey } from '@/utils/TranslationUtils';
import { useTranslation } from 'react-i18next';

export interface TempShareDropdownProps {
  handleUpload: (evt: React.SyntheticEvent<any>) => void;
  style?: React.CSSProperties;
  className?: string;

  // Allows temporarily replacing the content without
  // the need to refetch all temp share data
  overrideContent?: React.ReactElement | null;
}

interface DataProps {
  files: API.TempShareItem[];
}

type Props = TempShareDropdownProps & DataProps & DataProviderDecoratorChildProps;

const getDropdownItem = (
  file: API.TempShareItem,
  onClick: (p: API.TempShareItem) => void,
) => {
  let title = file.name;
  if (file.user) {
    title += ` (${file.user.nicks})`;
  }

  return (
    <MenuItemLink
      key={file.id}
      onClick={() => onClick(file)}
      icon={<FileIcon typeInfo={file.type} />}
    >
      {title}
    </MenuItemLink>
  );
};

const TempShareDropdown = React.memo<Props>(function TempShareDropdown({
  files,
  handleUpload,
  style,
  className,
  overrideContent,
  socket,
}) {
  const { t } = useTranslation();

  if (!!overrideContent) {
    return overrideContent;
  }

  const onClickFile = (file: API.TempShareItem) => {
    socket
      .delete(`${ShareConstants.TEMP_SHARES_URL}/${file.id}`)
      .then((res) => {
        NotificationActions.success({
          title: t(toI18nKey('fileNoLongerShared', UI.Modules.COMMON), {
            defaultValue: 'File {{file.name}} is no longer shared',
            replace: {
              file,
            },
          }),
        });
      })
      .catch((e) => {
        NotificationActions.apiError('Failed to remove shared item', e);
      });
  };

  const classNames = cx('top left pointing actions', className);

  return (
    <SectionedDropdown
      className={classNames}
      triggerIcon="plus"
      button={true}
      contextElement=".message-view"
    >
      <MenuItemLink onClick={handleUpload} icon={IconConstants.UPLOAD}>
        {translate('Send files', t, UI.Modules.COMMON)}
      </MenuItemLink>
      <MenuSection
        caption={translate('Remove shared file', t, UI.Modules.COMMON)}
        icon={IconConstants.FILE}
      >
        {files.map((p) => getDropdownItem(p, onClickFile))}
      </MenuSection>
    </SectionedDropdown>
  );
});

export default DataProviderDecorator<TempShareDropdownProps, DataProps>(
  TempShareDropdown,
  {
    urls: {
      files: ShareConstants.TEMP_SHARES_URL,
    },
    onSocketConnected: (addSocketListener, { refetchData }) => {
      addSocketListener(ShareConstants.MODULE_URL, ShareConstants.TEMP_ITEM_ADDED, () =>
        refetchData(),
      );
      addSocketListener(ShareConstants.MODULE_URL, ShareConstants.TEMP_ITEM_REMOVED, () =>
        refetchData(),
      );
    },
    loaderText: null,
  },
);
