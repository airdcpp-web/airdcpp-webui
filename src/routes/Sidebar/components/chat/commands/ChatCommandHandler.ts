import { actionAccess } from 'utils/ActionUtils';

import * as API from 'types/api';
import * as UI from 'types/ui';

const getHelpString = (
  commands: UI.ChatCommandList,
  session: UI.AuthenticatedSession,
) => {
  const commandHelp = Object.keys(commands)
    .filter((command) => actionAccess(commands[command], session))
    .map((command) => `\t/${command} - ${commands[command].help}`)
    .join('\n');

  return `
\tChat commands

${commandHelp}
`;
};

const CommandHandler = (chatController: UI.ChatController) => {
  const { chatCommands } = chatController;
  return {
    handle: (
      command: string,
      params: UI.ChatCommandParam,
      commandProps: UI.ChatCommandProps,
    ) => {
      if (chatCommands[command]) {
        chatCommands[command].handler(params, chatController, commandProps);
      } else if (command === 'help') {
        const { session, chatApi } = chatController;
        const text = getHelpString(chatCommands, commandProps.session);
        const message: API.OutgoingChatStatusMessage = {
          text,
          severity: API.SeverityEnum.INFO,
          type: API.StatusMessageTypeEnum.PRIVATE,
          owner: `session:${commandProps.session.sessionId}`,
        };

        chatApi.sendStatusMessage(commandProps.socket, session, message);
      }
    },
  };
};

export default CommandHandler;
