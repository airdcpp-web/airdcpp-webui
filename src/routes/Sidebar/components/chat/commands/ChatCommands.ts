import { runBackgroundSocketAction } from '@/utils/ActionUtils';

import * as API from '@/types/api';
import * as UI from '@/types/ui';

type ClearHandler = (id: API.IdType) => Promise<any>;

export const buildChatCommands = (
  editAccess: API.AccessEnum,
  clearAPIAction: ClearHandler,
) => {
  const handleMe: UI.ChatCommandHandler = (params, { chatApi, session }, { socket }) => {
    if (params) {
      chatApi.sendChatMessage(socket, session, params, true);
    }
  };

  const handleClear: UI.ChatCommandHandler = (params, { session }, { t }) => {
    runBackgroundSocketAction(() => clearAPIAction(session.id), t);
  };

  const commands: UI.ChatCommandList = {
    clear: {
      help: 'Clear message cache',
      handler: handleClear,
      access: editAccess,
    },
    me: {
      help: 'Send message in third person',
      handler: handleMe,
    },
  };

  return commands;
};
