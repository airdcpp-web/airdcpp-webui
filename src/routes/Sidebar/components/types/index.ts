//import * as React from 'react';

import { IconType } from 'components/semantic/Icon';

import * as API from 'types/api';
import * as UI from 'types/ui';

export interface SessionLayoutLabels {
  // Label for button that opens a new session
  newCaption?: React.ReactNode;

  // Label for button that opens a new session
  newDescription?: React.ReactNode;

  // Label for button that opens a new session
  newIcon?: IconType;
}

export interface SessionLayoutManageProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> {
  // Unique ID of the section (used for storing and loading the previously open tab)
  baseUrl: string;

  // Array of the items to list
  items: SessionT[];

  // Session actions to show in the action menu
  actionIds?: string[];

  uiActions: UI.ModuleActions<SessionT, void, UIActionT>;
  sessionApi: UI.SessionActions<SessionT> & SessionApiT;

  // Item ID that is currently active (if any)
  activeId: API.IdType | undefined;

  // AccessConstant defining whether the user has edit permission
  editAccess: API.AccessEnum;

  unreadInfoStore: UI.UnreadInfoStore;

  remoteMenuId: string;
}

interface SessionMainLayoutComponents {
  newButton: React.ReactElement<any> | null;
  listActionMenuGetter: () => React.ReactNode;
  itemHeaderTitle: React.ReactNode;
  itemHeaderDescription: React.ReactNode;
  itemHeaderIcon: IconType;
  sessionMenuItems: React.ReactNode[];
}

export interface SessionMainLayoutProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionsT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> extends Pick<
      SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>,
      'unreadInfoStore'
    >,
    SessionMainLayoutComponents {
  activeItem: SessionT | null;
  onKeyDown: (event: React.KeyboardEvent) => void;
  children: React.ReactNode;

  actions: SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>['uiActions'];
}

export type SessionChildProps<
  SessionT extends UI.SessionItemBase,
  SessionApiT extends object = UI.EmptyObject,
  UIActionsT extends UI.ActionListType<UI.SessionItemBase> = UI.EmptyObject,
> = Pick<
  SessionLayoutManageProps<SessionT, SessionApiT, UIActionsT>,
  'sessionApi' | 'uiActions'
> &
  UI.RouteComponentProps & {
    session: SessionT;
    sessionT: UI.ModuleTranslator;
  };

export interface NewSessionLayoutProps extends UI.RouteComponentProps {
  sessionT: UI.ModuleTranslator;
}

export interface SessionLocationState {
  pending?: boolean;
}
