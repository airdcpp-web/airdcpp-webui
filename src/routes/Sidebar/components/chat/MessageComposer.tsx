import * as React from 'react';
import classNames from 'classnames';

import { MentionsInput, Mention, OnChangeHandlerFunc, DataFunc } from 'react-mentions';
import Dropzone, { DropzoneRef } from 'react-dropzone';
import { RouteComponentProps, withRouter } from 'react-router-dom';

import { loadSessionProperty, saveSessionProperty, useMobileLayout } from 'utils/BrowserUtils';
import ChatCommandHandler from './ChatCommandHandler';
import NotificationActions from 'actions/NotificationActions';

import UserConstants from 'constants/UserConstants';
import SocketService from 'services/SocketService';
import { ChatLayoutProps } from 'routes/Sidebar/components/chat/ChatLayout';

import * as API from 'types/api';
import * as UI from 'types/ui';

import { ErrorResponse } from 'airdcpp-apisocket';
import FilePreviewDialog from './FilePreviewDialog';
import TempShareDropdown from './TempShareDropdown';
import { TFunction } from 'i18next';
import { translate, toI18nKey } from 'utils/TranslationUtils';
import { formatSize } from 'utils/ValueFormat';
import LoginStore from 'stores/LoginStore';
import AccessConstants from 'constants/AccessConstants';
import Icon from 'components/semantic/Icon';
import Button from 'components/semantic/Button';
import IconConstants from 'constants/IconConstants';


const getMentionFieldStyle = (mobileLayout: boolean) => {
  return {
    suggestions: {
      list: {
        width: 200,
        overflow: 'auto',
        position: 'absolute',
        bottom: 3,
        left: 17,
        backgroundColor: 'white',
        border: '1px solid rgba(0,0,0,0.15)',
        fontSize: 10,
      },

      item: {
        padding: '5px 15px',
        borderBottom: '1px solid rgba(0,0,0,0.15)',

        '&focused': {
          background: 'rgba(0,0,0,.03)',
          color: 'rgba(0,0,0,.95)',
        },
      },
    },
    highlighter: {
      minHeight: !mobileLayout ? 63 : 42,
      maxHeight: 200,
      margin: 0,
    },
    input: {
      minHeight: !mobileLayout ? 63 : 42,
      maxHeight: 200,
      margin: 0,

      // Disable weird hack from react-mentions which breaks styling on iOS
      // tslint:disable-next-line:max-line-length
      // https://github.com/signavio/react-mentions/blob/e2b0273d4a629af8107ba1738f0a5846ada790f1/src/MentionsInput.js#L1033
      marginLeft: 0,
      marginTop: 0,
    },
  };
};

export interface MessageComposerProps extends 
  Pick<ChatLayoutProps, 'chatApi' | 'session' | 'chatActions' | 'handleFileUpload'> {

  t: TFunction;
}

const getStorageKey = (props: RouteComponentProps) => {
  return `last_message_${props.location.pathname}`;
};

const loadState = (props: RouteComponentProps) => {
  return {
    text: loadSessionProperty(getStorageKey(props), ''),
  };
};


const saveState = (state: State, props: RouteComponentProps) => {
  saveSessionProperty(getStorageKey(props), state.text);
};

const userToMention = (user: API.HubUser) => {
  return {
    id: user.cid,
    display: user.nick,
  };
};

interface State {
  text: string;
  files: File[] | null;
  uploading: boolean;
}

class MessageComposer extends React.Component<MessageComposerProps & RouteComponentProps> {
  /*static propTypes = {
    // Actions for this chat session type
    actions: PropTypes.object.isRequired,
    session: PropTypes.object.isRequired,
  };*/

  dropzoneRef = React.createRef<DropzoneRef>();

  handleCommand = (text: string) => {
    const { location } = this.props;

    let command, params;

    {
      // Parse the command
      const whitespace = text.indexOf(' ');
      if (whitespace === -1) {
        command = text.substr(1);
      } else {
        command = text.substr(1, whitespace - 1);
        params = text.substr(whitespace + 1);
      }
    }

    ChatCommandHandler(this.props).handle(command, params, location);
  }

  handleSend = (text: string) => {
    const { chatApi, session } = this.props;
    chatApi.sendChatMessage(session, text);
  }

  componentWillUnmount() {
    saveState(this.state, this.props);
  }

  componentDidUpdate(prevProps: RouteComponentProps, prevState: State) {
    if (prevProps.location.pathname !== this.props.location.pathname) {
      saveState(prevState, prevProps);
      this.setState(loadState(this.props));
    }
  }

  handleChange: OnChangeHandlerFunc = (event, markupValue, plainValue) => {
    this.setState({ 
      text: plainValue 
    });
  }

  onKeyDown = (event: React.KeyboardEvent) => {
    if ((event.code === 'Enter' || event.code === 'NumpadEnter') && !event.shiftKey) {
      event.preventDefault();
      this.sendText();
    }
  }

  sendText = () => {
    // Trim only from end to allow chat messages such as " +help" to be
    // sent to other users
    // This will also prevent sending empty messages
    const text = this.state.text.replace(/\s+$/, '');

    if (text) {
      if (text[0] === '/') {
        this.handleCommand(text);
      }

      this.handleSend(text);
    }

    this.setState({ text: '' });
  }

  findUsers: DataFunc = (value, callback) => {
    const { session } = this.props;
    SocketService.post(UserConstants.SEARCH_NICKS_URL, { 
      pattern: value, 
      max_results: 5,
      hub_urls: session.hub_url ? [ session.hub_url ] : undefined,
    })
      .then((users: API.HubUser[]) => callback(users.map(userToMention)))
      .catch((error: ErrorResponse) => 
        console.log(`Failed to fetch suggestions: ${error}`)
      );
  }
  
  appendText = (text: string) => {
    let newText = this.state.text;
    if (newText) {
      newText += ' ';
    }
    newText += text;

    this.setState({
      text: newText
    });
  }

  onDropFile = (acceptedFiles: File[]) => {
    const { t } = this.props;
    const maxSize = 100 * 1024 * 1024;

    const files = acceptedFiles
      .filter(file => {
        if (file.size > maxSize) {
          NotificationActions.error({
            title: file.name,
            message: t(toI18nKey('fileTooLarge', UI.Modules.COMMON), {
              defaultValue: 'File is too large (maximum size is {{maxSize}})',
              replace: {
                maxSize: formatSize(maxSize, t)
              }
            })
          });
    
          return false;
        }

        return true;
      });

    if (!files.length) {
      return;
    }

    this.setState({
      files
    });
  }

  onUploadFiles = async (files: File[]) => {
    this.resetFiles();
    this.setState({
      uploading: true
    });

    const { handleFileUpload } = this.props;
    for (let file of files) {
      try {
        const res = await handleFileUpload(file);
        this.appendText(res.magnet);
      } catch (e) {
        NotificationActions.apiError('Failed to upload the file', e);
      }
    }

    this.setState({
      uploading: false
    });
  }

  onPaste = (evt: React.ClipboardEvent<HTMLTextAreaElement>) => {
    if (evt.clipboardData && evt.clipboardData.files && (evt.clipboardData as any).files.length) {
      let files: File[] = [];

      {
        // DataTransferItemList isn't a normal array, convert it first
        // https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItemList
        const dataTransferItemList = (evt.clipboardData as any).files as any[];
        for (let i = 0; i < dataTransferItemList.length; i++) {
          files.push(dataTransferItemList[i]);
        }
      }

      evt.preventDefault();
      this.onDropFile(files);
    }
  }

  resetFiles = () => {
    this.setState({
      files: null,
    });
  }

  state: State = {
    ...loadState(this.props),
    files: null,
    uploading: false,
  };

  render() {
    const mobile = useMobileLayout();
    const className = classNames(
      'ui form composer',
      { small: mobile },
      { large: !mobile },
    );

    const hasFileUploadAccess = LoginStore.hasAccess(AccessConstants.FILESYSTEM_EDIT) && 
      LoginStore.hasAccess(AccessConstants.SETTINGS_EDIT);

    const { files, text, uploading } = this.state;
    const { t } = this.props;

    const sendButton = (
      <Button 
        className="blue large icon send" 
        onClick={ this.sendText }
        caption={ <Icon icon={ IconConstants.SEND }/> }
        loading={ uploading }
      />
    );

    return (
      <>
        { !!files && (
          <FilePreviewDialog
            files={ files }
            onConfirm={ this.onUploadFiles }
            onReject={ this.resetFiles }
            title={ translate('Send files', t, UI.Modules.COMMON) }
          />
        ) }
        <Dropzone
          onDrop={ this.onDropFile }
          ref={ this.dropzoneRef }
          disabled={ !hasFileUploadAccess }
          multiple={ true }
          minSize={ 1 }
          noClick={ true }
          // Handle max size check elsewhere (report errors)
        >
          {({ getRootProps, getInputProps, open }) => (
            <div 
              className={ className }
              { ...getRootProps() }
            >
              <input 
                { ...getInputProps() }
              />
              <MentionsInput 
                className="input"
                value={ text } 
                onChange={ this.handleChange }
                onKeyDown={ this.onKeyDown }
                style={ getMentionFieldStyle(mobile) }
                autoFocus={ !mobile }
                onPaste={ this.onPaste }
              >
                <Mention 
                  trigger="@"
                  data={ this.findUsers }
                  appendSpaceOnAdd={ false }
                />
              </MentionsInput>
              { !hasFileUploadAccess || uploading ? sendButton : (
                <TempShareDropdown
                  className="blue large" 
                  handleUpload={ open }
                  overrideContent={ !text ? null : sendButton }
                />
              ) }
            </div>
          ) }
        </Dropzone>
      </>
    );
  }
}

const Decorated = withRouter(MessageComposer);

export default Decorated;