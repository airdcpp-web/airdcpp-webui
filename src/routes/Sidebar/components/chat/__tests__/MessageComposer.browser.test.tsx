import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';

import { renderDataNode } from '@/tests/render/test-renderers';

import { initCommonDataMocks } from '@/tests/mocks/mock-data-common';

import * as API from '@/types/api';

import { getMockServer, MockServer } from '@/tests/mocks/mock-server';
import { MessageComposer } from '../MessageComposer';
import { useTranslation } from 'react-i18next';
import { buildChatCommands } from '../commands/ChatCommands';
import { VIEW_FIXED_HEIGHT } from '@/tests/render/test-containers';
import { waitFor } from '@testing-library/dom';
import { setupUserEvent } from '@/tests/helpers/test-form-helpers';
import { installTempShareMocks } from '@/tests/mocks/mock-share';
import {
  clickButton,
  clickMenuItem,
  expectRequestToMatchSnapshot,
  waitForLoader,
} from '@/tests/helpers/test-helpers';
import { TempShareItem1 } from '@/tests/mocks/api/share';
import { formatTempShareDropdownItem } from '../TempShareDropdown';
import ShareConstants from '@/constants/ShareConstants';

describe('Message composer', () => {
  let server: MockServer;
  const getSocket = async () => {
    const commonData = await initCommonDataMocks(server);

    installTempShareMocks(server);

    return { commonData, server };
  };

  const renderLayout = async (settings: object = {}) => {
    const onUpload = vi.fn();

    const mockChatApi = {
      sendChatMessage: vi.fn(),
      sendStatusMessage: vi.fn(),
      fetchMessages: vi.fn(),

      setRead: vi.fn(),
      removeSession: vi.fn(),
    };

    const mockChatSession = {
      id: 'test-session',
      read: false,
    };

    const onClear = vi.fn(() => Promise.resolve());
    const mockChatCommands = buildChatCommands(API.AccessEnum.PRIVATE_CHAT_EDIT, onClear);

    const { commonData, server, ...other } = await getSocket();

    const MessageComposerLayoutTest = () => {
      const { t } = useTranslation();
      return (
        <MessageComposer
          t={t}
          handleFileUpload={onUpload}
          chatApi={mockChatApi}
          chatSession={mockChatSession}
          chatCommands={mockChatCommands}
          sessionType="mock_session_type"
        />
      );
    };

    const renderData = renderDataNode(
      <MessageComposerLayoutTest />,
      commonData,
      VIEW_FIXED_HEIGHT,
    );

    return {
      ...renderData,
      ...other,
      ...commonData,
      mockChatApi,
      mockChatSession,
      mockChatCommands,
      onClear,
      onUpload,
    };
  };

  beforeEach(() => {
    server = getMockServer();
  });

  afterEach(() => {
    server.stop();
  });

  describe('Chat messages', () => {
    test('normal', async () => {
      const { getByRole, mockChatApi } = await renderLayout();

      await waitFor(() => expect(getByRole('textbox')).toBeInTheDocument());

      // Use the send button
      const userEvent = setupUserEvent();
      const messageInput = getByRole('textbox');
      await userEvent.type(messageInput, 'Test message 1');
      await waitFor(() => getByRole('button', { name: 'Send message' }));
      await userEvent.click(getByRole('button', { name: 'Send message' }));
      await waitFor(() => {
        expect(messageInput).toHaveValue('');
      });

      expect(mockChatApi.sendChatMessage).toHaveBeenCalledTimes(1);
      expect(mockChatApi.sendChatMessage.mock.calls[0][2]).toMatchSnapshot();

      // Use the Enter key to send
      await userEvent.type(messageInput, 'Test message 2');
      await userEvent.keyboard('{Enter}');
      await waitFor(() => {
        expect(messageInput).toHaveValue('');
      });
      expect(mockChatApi.sendChatMessage).toHaveBeenCalledTimes(2);
    });

    test('third person', async () => {
      const { getByRole, mockChatApi } = await renderLayout();

      await waitFor(() => expect(getByRole('textbox')).toBeInTheDocument());

      const userEvent = setupUserEvent();
      const messageInput = getByRole('textbox');
      await userEvent.type(messageInput, '/me Test message third person');
      await userEvent.keyboard('{Enter}');
      await waitFor(() => {
        expect(messageInput).toHaveValue('');
      });
      await waitFor(() => {
        expect(mockChatApi.sendChatMessage).toHaveBeenCalledTimes(2);
      });

      expect(mockChatApi.sendChatMessage.mock.calls[0][2]).toMatchSnapshot();
      expect(mockChatApi.sendChatMessage.mock.calls[1][2]).toMatchSnapshot();
    });
  });

  describe('Chat commands', () => {
    test('should handle help', async () => {
      const { getByRole, mockChatApi } = await renderLayout();

      await waitFor(() => expect(getByRole('textbox')).toBeInTheDocument());

      // Use the send button
      const userEvent = setupUserEvent();
      const messageInput = getByRole('textbox');
      await userEvent.type(messageInput, '/help');
      await userEvent.keyboard('{Enter}');

      expect(mockChatApi.sendChatMessage).toHaveBeenCalledTimes(1);
      expect(mockChatApi.sendStatusMessage).toHaveBeenCalledTimes(1);
      expect(mockChatApi.sendStatusMessage.mock.calls[0][2]).toMatchSnapshot();
    });

    test('should handle clear', async () => {
      const { getByRole, onClear } = await renderLayout();

      await waitFor(() => expect(getByRole('textbox')).toBeInTheDocument());

      // Use the send button
      const userEvent = setupUserEvent();
      const messageInput = getByRole('textbox');
      await userEvent.type(messageInput, '/clear');
      await userEvent.keyboard('{Enter}');

      expect(onClear).toHaveBeenCalledTimes(1);
    });
  });

  describe('Temp share', () => {
    test('should add shared files', async () => {
      const { getByTestId, getByRole, findByText, onUpload } = await renderLayout();

      await waitFor(() => expect(getByRole('textbox')).toBeInTheDocument());

      const file = new File(['test'], 'test.txt', { type: 'text/plain' });
      const input = getByTestId('dropzone');

      const userEvent = setupUserEvent();
      await userEvent.upload(input, file);

      await waitFor(() => expect(findByText('test.txt')).toBeTruthy());

      clickButton('Upload', getByRole);

      await waitFor(() => expect(onUpload).toHaveBeenCalledTimes(1));
      expect(onUpload.mock.calls[0]).toMatchInlineSnapshot(`
        [
          File {
            "path": "./test.txt",
            "relativePath": "./test.txt",
          },
        ]
      `);
    });

    test('should remove shared files', async () => {
      const { getByLabelText, queryByRole, getByRole } = await renderLayout();

      await waitForLoader(queryByRole);

      const userEvent = setupUserEvent();
      const dropdown = getByLabelText('Temp share');
      await userEvent.click(dropdown);

      const onItemRemoved = vi.fn();
      server.addRequestHandler(
        'DELETE',
        `${ShareConstants.TEMP_SHARES_URL}/${TempShareItem1.id}`,
        undefined,
        onItemRemoved,
      );

      await clickMenuItem(formatTempShareDropdownItem(TempShareItem1), getByRole);

      await waitFor(() => expect(onItemRemoved).toHaveBeenCalledTimes(1));
      expectRequestToMatchSnapshot(onItemRemoved);
    });
  });
});
