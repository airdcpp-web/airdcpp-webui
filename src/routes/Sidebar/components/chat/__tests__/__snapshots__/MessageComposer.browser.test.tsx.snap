// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Message composer > Chat commands > should handle help 1`] = `
{
  "owner": "session:4",
  "severity": "info",
  "text": "
	Chat commands

	/clear - Clear message cache
	/me - Send message in third person
",
  "type": "private",
}
`;

exports[`Message composer > Chat messages > normal 1`] = `
{
  "text": "Test message 1",
}
`;

exports[`Message composer > Chat messages > third person 1`] = `
{
  "text": "Test message third person",
  "third_person": true,
}
`;

exports[`Message composer > Chat messages > third person 2`] = `
{
  "text": "/me Test message third person",
}
`;
